swap/src/cli/cancel_and_refund.rs

```diff
@@ -55,9 +56,22 @@ pub async fn cancel(
 
     tracing::info!(%swap_id, "Manually cancelling swap");
 
-    let (txid, subscription) = match state6.submit_tx_cancel(bitcoin_wallet.as_ref()).await {
-        Ok(txid) => txid,
+    match state6.submit_tx_cancel(bitcoin_wallet.as_ref()).await {
+        Ok((txid, _)) => {
+            let state = BobState::BtcCancelled(state6);
+            db.insert_latest_state(swap_id, state.clone().into())
+                .await?;
+            Ok((txid, state))
+        }
         Err(err) => {
+            if let Ok(tx) = state6.check_for_tx_cancel(bitcoin_wallet.as_ref()).await {
+                // Alice already cancelled swap, so Bob sets his state to BtcCancelled.
+                let state = BobState::BtcCancelled(state6);
+                db.insert_latest_state(swap_id, state.clone().into())
+                    .await?;
+                tracing::info!("Cancel transaction has already been confirmed on chain. The swap has therefore already been cancelled by Alice.");
+                return Ok((tx.txid(), state));
+            }
             if let Ok(error_code) = parse_rpc_error_code(&err) {
```

Comment:
We have only covered this cases here: Cancel timelock has expired but cancel transaction has already been published by Alice

We still need to cover the case where we try to publish the cancel transaction but can't because the timelock has not expired yet. We have to print a fitting log message in that case.

Reply:
I added check and message for this case.

Reply:
Should we still throw an unappealing error if the cancel time has not yet expired or return original state? This will break tests.